### ISSUE TITLE
Old Che (5.2.2) IDE extension to new Che extension

### ISSUE THREAD

**Issue by vnea** [eclipse-che/che#5032]
Hello,

I developed a Che 5.2.2 IDE extension and I want to report this extension to the lastest version of Che (currently 5.9.1). I am using **Windows 10** and **Docker Toolbox**.

Before trying to report my extension, I checked out the [docs](https://www.eclipse.org/che/docs/assemblies/intro/index.html) and I noticed it has changed since the last time I read it (end of February) .

If I understand correctly, we now have to use the **archetype** tool to generate a predifined extension (plugin-menu-archetype, plugin-embedjs-archetype, etc.) depending of what the extension will do and then just adapt the code before building and running it.

To generate a predefined extension, I tried the following command (from [here](https://www.eclipse.org/che/docs/assemblies/assembly-lifecycle/index.html#generate)) with **eclipse/che-cli:5.9.1**:

```
 docker run -it --rm 
     -v /var/run/docker.sock:/var/run/docker.sock
     -v /c/dev/che/archetype:/archetype
     -v /c/dev/che/chedata_5.9.1:/data
     -v /c/Users/victorn/.m2/repository:/m2
     eclipse/che-cli:5.9.1
         archetype generate
```
and then:

> INFO: (che archetype): Welcome to Eclipse Che custom assembly generator!
> INFO: (che archetype):
> INFO: (che archetype): You can skip this message with '--no:interactive'.
> INFO: (che archetype):
> INFO: (che archetype): This generator requires:
> INFO: (che archetype):   1. Maven 3.3+ to be installed on your host.
> INFO: (che archetype):   2. Your host Maven M2 repo mounted to ':/m2'.
> INFO: (che archetype):   3. A local path for us to place the assembly mounted to ':/archetype'.
> 
>  **Ready? [Y/n] y**
> 1\) agent-archetype                - Assembly with sample agent
> 2) plugin-menu-archetype          - Assembly with IDE extension to customize menu
> 3) plugin-wizard-archetype        - Assembly with custom C project type extension
> 4) plugin-serverservice-archetype - Assembly with simple IDE extesion and a server service
> 5) plugin-embedjs-archetype       - Assembly with simple IDE extesion for using native javascript in widgets
> 6) plugin-json-archetype          - Assembly with sample JSON project type, editor codeassistant, and workspace services
> **Please enter your choice: 5**

Finaly I got the following error:

> [INFO] ------------------------------------------------------------------------
> [INFO] Building Maven Stub Project (No POM) 1
> [INFO] ------------------------------------------------------------------------
> [INFO]
> [INFO] >>> maven-archetype-plugin:2.4:generate (default-cli) > generate-sources @ standalone-pom >>>
> [INFO]
> [INFO] <<< maven-archetype-plugin:2.4:generate (default-cli) < generate-sources @ standalone-pom <<<
> [INFO]
> [INFO] --- maven-archetype-plugin:2.4:generate (default-cli) @ standalone-pom ---
> [INFO] Generating project in Batch mode
> [INFO] Archetype defined by properties
> Downloading: http://maven.codenvycorp.com/content/groups/public/org/eclipse/che/archetype/plugin-embedjs-archetype/5.9.1/plugin-embedjs-archetype-5.9.1.jar
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time: 9.120 s
> [INFO] Finished at: 2017-05-10T09:24:13+00:00
> [INFO] Final Memory: 15M/162M
> [INFO] ------------------------------------------------------------------------
> **[ERROR] Failed to execute goal org.apache.maven.plugins:maven-archetype-plugin:2.4:generate (default-cli) on project standalone-pom: The desired archetype does not exist (org.eclipse.che.archetype:plugin-embedjs-archetype:5.9.1) -> [Help 1][ERROR]**
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.

As it failed, I tried the same command with **eclipse/che-cli:5.9.0**:
```
 docker run -it --rm 
     -v /var/run/docker.sock:/var/run/docker.sock
     -v /c/dev/che/archetype:/archetype
     -v /c/dev/che/chedata_5.9.0:/data
     -v /c/Users/victorn/.m2/repository:/m2
     eclipse/che-cli:5.9.0
         archetype generate
```
and then I typed **y** then **5** and I got the following error:

> [INFO] ------------------------------------------------------------------------
> [INFO] Building Maven Stub Project (No POM) 1
> [INFO] ------------------------------------------------------------------------
> [INFO]
> [INFO] >>> maven-archetype-plugin:2.4:generate (default-cli) > generate-sources @ standalone-pom >>>
> [INFO]
> [INFO] <<< maven-archetype-plugin:2.4:generate (default-cli) < generate-sources @ standalone-pom <<<
> [INFO]
> [INFO] --- maven-archetype-plugin:2.4:generate (default-cli) @ standalone-pom ---
> [INFO] Generating project in Batch mode
> [INFO] Archetype defined by properties
> [INFO] ----------------------------------------------------------------------------
> [INFO] Using following parameters for creating project from Archetype: plugin-embedjs-archetype:5.9.0
> [INFO] ----------------------------------------------------------------------------
> [INFO] Parameter: groupId, Value: com.sample
> [INFO] Parameter: artifactId, Value: assembly
> [INFO] Parameter: version, Value: 5.9.0
> [INFO] Parameter: package, Value: com.sample
> [INFO] Parameter: packageInPathFormat, Value: com/sample
> [INFO] Parameter: package, Value: com.sample
> [INFO] Parameter: version, Value: 5.9.0
> [INFO] Parameter: groupId, Value: com.sample
> [INFO] Parameter: artifactId, Value: assembly
> **[WARNING] The directory /home/user/che-build/assembly already exists.**
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time: 4.178 s
> [INFO] Finished at: 2017-05-10T09:34:40+00:00
> [INFO] Final Memory: 12M/162M
> [INFO] ------------------------------------------------------------------------
> **[ERROR] Failed to execute goal org.apache.maven.plugins:maven-archetype-plugin:2.4:generate (default-cli) on project standalone-pom: org.apache.maven.archetype.exception.ArchetypeGenerationFailure: Error merging velocity templates: /home/user/che-build/assembly/pom.xml (Permission denied) -> [Help 1]**
> [ERROR]
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.

I noticed the following line:
> [WARNING] The directory /home/user/che-build/assembly already exists.

Is it possible that there is a conflict with my old Che 5.2.2 build ?
I checked if the folder exists with:
`docker-machine ssh`
and then:
`cd /home/user/che-build`
but the folder does not exist (`-sh: cd: can't cd to /home/user/che-build`)...

I then tried with **eclipse/che-cli:5.8.0** and **eclipse/che-cli:5.7.0** and got this error in both cases:
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-archetype-plugin:2.4:generate (default-cli) on project standalone-pom: The desired archetype does not exist (org.eclipse.che.archetype:plugin-embedjs-archetype:5.6.0-SNAPSHOT) -> [Help 1]

**Comment by vnea** [comment]
I rebooted my computer and re-executed the command:
```
 docker run -it --rm 
     -v /var/run/docker.sock:/var/run/docker.sock
     -v /c/dev/che/archetype:/archetype
     -v /c/dev/che/chedata_5.9.0:/data
     -v /c/Users/victorn/.m2/repository:/m2
     eclipse/che-cli:5.9.0
         archetype generate
```
but now I have the error:

> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-archetype-plugin:2.4:generate (default-cli) on project standalone-pom: org.apache.maven.archetype.exception.ArchetypeGenerationFailure: Error merging velocity templates: /home/user/che-build/assembly/pom.xml (**No such file or directory**) -> [Help 1]

instead of

> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-archetype-plugin:2.4:generate (default-cli) on project standalone-pom: org.apache.maven.archetype.exception.ArchetypeGenerationFailure: Error merging velocity templates: /home/user/che-build/assembly/pom.xml (**Permission denied**) -> [Help 1]

The warning

> [WARNING] The directory /home/user/che-build/assembly already exists.

also disappeared.

Am I doing something wrong ?

**Comment by vnea** [comment]
I finaly found out why it wasn't working.

Just after running the command:
```
 docker run -it --rm 
     -v /var/run/docker.sock:/var/run/docker.sock
     -v /c/dev/che/archetype:/archetype
     -v /c/dev/che/chedata_5.9.0:/data
     -v /c/Users/victorn/.m2/repository:/m2
     eclipse/che-cli:5.9.0
         archetype generate
```
I saw the following message:

> WARN: Boot2docker detected - ensure :/data is mounted to %userprofile%

So I changed the paths of archetype and data:
```
 docker run -it --rm 
     -v /var/run/docker.sock:/var/run/docker.sock
     -v /c/Users/victorn/che_5.9.0/archetype:/archetype
     -v /c/Users/victorn/che_5.9.0/tmp:/data
     -v /c/Users/victorn/.m2/repository:/m2
     eclipse/che-cli:5.9.0
         archetype generate
```
and the generation worked !

I think you should add this in your documentation because it is currently written 

` docker run -it --rm -v /c/tmp:/data ...`

and it does not work if **Boot2docker** is used (maybe with other tools it would not work too).
<br>

Then I wanted to build with:
```
 docker run -it --rm 
     -v /var/run/docker.sock:/var/run/docker.sock
     -v /c/Users/victorn/che_5.9.0/archetype:/archetype
     -v /c/Users/victorn/che_5.9.0/tmp:/data
     -v /c/Users/victorn/.m2/repository:/m2
     eclipse/che-cli:5.9.0
         archetype build
```
but I got this error:

> [INFO] ------------------------------------------------------------------------
> [INFO] Building Plugin Embed JavaScript :: Che Assembly :: Assembly Parent 5.9.0
> [INFO] ------------------------------------------------------------------------
> **[WARNING] The POM for org.eclipse.che:assembly-ide-war:jar:classes:5.9.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details**
> **[WARNING] The POM for org.eclipse.che:assembly-ide-war:war:5.9.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details**
> [INFO]
> [INFO] --- maven-clean-plugin:2.6.1:clean (default-clean) @ assembly-ide-war ---
> [INFO]
> [INFO] --- build-helper-maven-plugin:1.9.1:parse-version (parse-version) @ assembly-ide-war ---
> [INFO]
> [INFO] --- buildnumber-maven-plugin:1.3:create (default) @ assembly-ide-war ---
> [INFO] ShortRevision tag detected. The value is '16'.
> [INFO] Executing: /bin/sh -c cd /home/user/che-build/assembly-che/assembly-ide-war && git rev-parse --verify --short=16 HEAD
> [INFO] Working directory: /home/user/che-build/assembly-che/assembly-ide-war
> [INFO] Storing buildNumber: no_revision at timestamp:  2017-05-10 16:26:50
> **[WARNING] Cannot get the branch information from the git repository:**
> **Detecting the current branch failed: fatal: Not a git repository (or any parent up to mount point /home/user/che-build)**
> Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
> 
> [INFO] ShortRevision tag detected. The value is '16'.
> [INFO] Executing: /bin/sh -c cd /home/user/che-build/assembly-che/assembly-ide-war && git rev-parse --verify --short=16 HEAD
> [INFO] Working directory: /home/user/che-build/assembly-che/assembly-ide-war
> [INFO] Storing buildScmBranch: UNKNOWN_BRANCH
> [INFO]
> [INFO] --- maven-enforcer-plugin:1.4:enforce (enforce-java-version) @ assembly-ide-war ---
> [INFO]
> [INFO] --- maven-enforcer-plugin:1.4:enforce (enforce-maven-version) @ assembly-ide-war ---
> [INFO]
> [INFO] --- maven-enforcer-plugin:1.4:enforce (distribution-management-used) @ assembly-ide-war ---
> [INFO]
> [INFO] --- maven-enforcer-plugin:1.4:enforce (no-distribution-management-at-all) @ assembly-ide-war ---
> [INFO]
> [INFO] --- maven-enforcer-plugin:1.4:enforce (dependency) @ assembly-ide-war ---
> [WARNING] Invalid POM for org.eclipse.che:assembly-ide-war:jar:classes:5.9.0, transitive dependencies (if any) will not be available, enable debug logging for more details
> **[WARNING] Invalid POM for org.eclipse.che:assembly-ide-war:war:5.9.0, transitive dependencies (if any) will not be available, enable debug logging for more details**
> [INFO]
> [INFO] --- maven-enforcer-plugin:1.4:enforce (ban-bad-dependencies) @ assembly-ide-war ---
> [INFO]
> [INFO] --- maven-sortpom-plugin:2.3.0:verify (default) @ assembly-ide-war ---
> [INFO] Skipping Sortpom
> [INFO]
> [INFO] --- license-maven-plugin:2.10:check (check-headers) @ assembly-ide-war ---
> [INFO] Checking licenses...
> [INFO]
> [INFO] --- exec-maven-plugin:1.4.0:java (gwt-xml) @ assembly-ide-war ---
> **[WARNING]**
> **java.lang.ClassNotFoundException: org.eclipse.che.util.GwtXmlGenerator**
>         **at java.net.URLClassLoader.findClass(URLClassLoader.java:381)**
>         **at java.lang.ClassLoader.loadClass(ClassLoader.java:424)**
>         **at java.lang.ClassLoader.loadClass(ClassLoader.java:357)**
>         **at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:281)**
>         **at java.lang.Thread.run(Thread.java:745)**
> [INFO] ------------------------------------------------------------------------
> [INFO] Reactor Summary:
> [INFO]
> [INFO] Plugin Embed JavaScript :: Parent pom .............. SUCCESS [ 31.907 s]
> [INFO] Plugin Embed JavaScript :: Plugins ::  Parent ...... SUCCESS [  0.164 s]
> [INFO] Plugin Embed JavaScript :: Plugin Embed JavaScript :: Parent SUCCESS [  0.199 s]
> [INFO] Plugin Embed JavaScript :: Plugin Embed JavaScript :: IDE SUCCESS [01:26 min]
> [INFO] Plugin Embed JavaScript :: Che Assembly :: Assembly Parent SUCCESS [  0.162 s]
> [INFO] Plugin Embed JavaScript :: Che Assembly :: Assembly Parent FAILURE [  6.535 s]
> [INFO] Plugin Embed JavaScript :: Che Assembly :: Assembly Parent SKIPPED
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time: 02:11 min
> [INFO] Finished at: 2017-05-10T16:26:55+00:00
> [INFO] Final Memory: 64M/988M
> [INFO] ------------------------------------------------------------------------
> **[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:java (gwt-xml) on project assembly-ide-war: An exception occured while executing the Java class. org.eclipse.che.util.GwtXmlGenerator -> [Help 1]**

Apparently there is a problem with the class **org.eclipse.che.util.GwtXmlGenerator**.

<br>
This morning I saw that Che 5.10.0 has been released and I tried to execute with this version all commands (docker run... archetype generate/build/run) and everything worked fine !

<br>
Now I will try to report all my Che 5.2.2 extensions to Che 5.10.0 and if it works, I will close the issue.

**Comment by ghost** [comment]
Yes, when using boot2docker, mounting userprofile is a must since in Docker for Windows the entire disc C gets mounted into a VM, unlike VirtualBox where only c/users/UserName is mounted.

**Comment by vnea** [comment]
My problem is solved, check the answers here #5054.
Thank you.
